toolchain/check/impl_lookup.cpp

```diff
@@ -575,91 +569,68 @@ static auto CollectCandidateImplsForQuery(
-    }
+  auto facet_lookup_result = LookupImplWitnessInSelfFacetValue(
+      context, non_canonical_query_self_inst_id, query_specific_interface);
+  if (facet_lookup_result.has_concrete_value()) {
+    return facet_lookup_result;
+  }
 
-    // NOTE: GetWitnessIdForImpl() does deduction, which can cause new impls
-    // to be imported, invalidating any pointer into `context.impls()`.
-    auto result = GetWitnessIdForImpl(
-        context, loc_id, query_is_concrete, query_self_const_id,
-        query_specific_interface, candidate.impl_id);
-    if (result.has_value()) {
-      return result;
+  // If the self type is a facet that provides a witness, then we are in an
+  // `interface` or an `impl`. In both cases, we don't want to do any impl
+  // lookups. The query will eventually resolve to a concrete witness when it
+  // can get it from the self facet value, when it has a specific applied in the
+  // future.
+  //
+  // In particular, this avoids a LookupImplWitness instruction in the eval
+  // block of an impl declaration from doing impl lookup. Specifically the
+  // lookup of the implicit .Self in `impl ... where .X`. If it does impl lookup
+  // when the eval block is run, it finds the same `impl`, tries to build a
+  // specific from it, which runs the eval block, creating a recursive loop that
+  // crashes.
+  bool self_facet_provides_witness = facet_lookup_result.has_value();
+  if (self_facet_provides_witness) {
+    if (auto bind = context.insts().TryGetAs<SemIR::BindSymbolicName>(
+            eval_query.query_self_inst_id)) {
+      const auto& entity = context.entity_names().Get(bind->entity_name_id);
+      if (entity.name_id == SemIR::NameId::PeriodSelf ||
+          entity.name_id == SemIR::NameId::SelfType) {
+        return EvalImplLookupResult::MakeNonFinal();
+      }
     }
   }
-  return EvalImplLookupResult::MakeNone();
-}
 
-auto LookupFinalImplWitnessForSpecificInterface(
-    Context& context, SemIR::LocId loc_id,
-    SemIR::ConstantId query_self_const_id,
-    SemIR::SpecificInterface query_specific_interface) -> SemIR::InstId {
-  // This would mean we need to UnwrapFacetAccessType(query_self_const_id), but
-  // it's already done by member access, which is the one use of this function.
-  CARBON_DCHECK(!context.insts().Is<SemIR::FacetAccessType>(
-      context.constant_values().GetInstId(query_self_const_id)));
+  SemIR::ConstantId query_self_const_id =
+      context.constant_values().Get(eval_query.query_self_inst_id);
 
   auto query_type_structure = BuildTypeStructure(
       context, context.constant_values().GetInstId(query_self_const_id),
       query_specific_interface);
   bool query_is_concrete =
       QueryIsConcrete(context, query_self_const_id, query_specific_interface);
 
+  // If we have a symbolic witness in the self query, then the query can not be
+  // concrete: the query includes a symbolic self value.
+  CARBON_DCHECK(!self_facet_provides_witness || !query_is_concrete);
```

Comment:
Usually we use CARBON_CHECK instead of DCHECK unless we have a specific performance concern.
```suggestion
  CARBON_CHECK(!self_facet_provides_witness || !query_is_concrete);
```